Password reset 3

shinwachi opened the pull request:
### Please give a short summary of what's included in this submission

Added password reset mechanism.  See: https://github.com/2018swecapstone/h2ms/wiki/Testing-password-reset-back-end-mechanism to understand how to use it.

### Checklist
- [x] The Pull Request title describes the changes I've made
- [x] I've reviewed the **Files changed** tab
- [x] The code in the **Files changed** tab meets our style guides
- [x] New classes and public methods in the **Files changed** tab are documented
- [x] I've added two reviewers for this pull request

stbenjam commented on src/main/java/edu/harvard/h2ms/service/EmailServiceImpl.java:
```diff
@@ -0,0 +1,33 @@
+
+@Component
+@Service("emailService")
+@PropertySource("classpath:mailserver.properties")
```
The app will start if you make this optional:

```java
@PropertySource(value = "classpath:mailserver.properties", ignoreResourceNotFound = true)
```

stbenjam commented:
Hm, even if you get it to ignore the missing properties the app refused to start for me because the configuration was then missing:

Bean method 'mailSender' not loaded because AnyNestedCondition 0 matched 2 did not; NestedCondition on MailSenderAutoConfiguration.MailSenderCondition.JndiNameProperty @ConditionalOnProperty (spring.mail.jndi-name) did not find property 'jndi-name'; NestedCondition on MailSenderAutoConfiguration.MailSenderCondition.HostProperty @ConditionalOnProperty (spring.mail.host) did not find property 'host'


What if you just ship the template as the properties file, and we override the settings with environment variables in the Jenkins deployment of the app?

shinwachi commented:
Thanks.  Inspired by Stephen's comment, I've looked up this one: https://blog.jayway.com/2014/02/16/spring-propertysource/

Which allows us to have alternative property files in a list format, like this: 
@PropertySources({
    @PropertySource("default.properties"),
    @PropertySource("overriding.properties")
})

I've set it to use the template.  It compiles, but still have issues down the line.  Will investigate and update the branch tonight.

shinwachi commented:
Ok, I've updated the branch with Stephen's suggestion of using mock email server (greeenmail), as well as override of property source if the main email setup file doesn't exist.

stbenjam commented on src/main/java/edu/harvard/h2ms/H2MSRestAppInitializer.java:
```diff
@@ -1,26 +1,26 @@
+	public static void main(String[] args) {
+		GreenMail greenMail = new GreenMail(ServerSetupTest.SMTP);
+        greenMail.start();
+		SpringApplication.run(H2MSRestAppInitializer.class, args);
```
Not sure if you looked at this, but having this in main doesn't strike me as very Spring-like, could it be a component instead? http://blog.netgloo.com/2014/11/13/run-code-at-spring-boot-startup/

That'd let us restrict it to run or not with @Profile annotations, for example.

shinwachi commented:
Ok, changed greenmail startup to be associated directly with email server related class.  Also, modified it so it won't start up without spring property value called emailServerType equal to greenmail.

stbenjam commented:
Oh neat, that's pretty clever.  This looks good to me, I'll try it out tomorrow.

shinwachi commented:
Thanks for the link.  Helped a lot.

stbenjam commented:
Looks like you and @MichaelDelaney were editing the same files, this needs rebased on his changes that were just merged.  

But code looks good :+1:

shinwachi commented:
Ok, merged.  I just realized that we made UserServiceImpl a repository of questions?  Shouldn't there be a different entity?  Anyway, password reset's done.

stbenjam commented:
It's not a question repository, it has the question repository autowired into it so @MichaelDelaney can get the compliance percentage for a user.

benjenkinsv95 commented on src/main/java/edu/harvard/h2ms/domain/core/User.java:
```diff
@@ -188,8 +206,38 @@ public boolean isCredentialsNonExpired() {
     @Override
     public boolean isEnabled() {
-        return true;
+        return true; //enabled;
```
I'm not sure this comment adds any value.

benjenkinsv95 commented on src/main/java/edu/harvard/h2ms/service/EmailServiceImpl.java:
```diff
@@ -0,0 +1,67 @@
+	public void sendEmail(SimpleMailMessage email) {
+		
+		
+		log.info("sending email to: "+email);
```
Awkward whitespace.  Might want to run an autoformatter on this file.

benjenkinsv95 commented on src/main/java/edu/harvard/h2ms/web/controller/PasswordController.java:
```diff
@@ -0,0 +1,132 @@
+		
+		
+		// fail silently to prevent probing
+		return new ResponseEntity<Object>(entity, HttpStatus.OK);
```
The comment seems awkward. Since the code below happens on success and failures.

benjenkinsv95 commented on src/main/java/edu/harvard/h2ms/web/controller/PasswordController.java:
```diff
@@ -0,0 +1,132 @@
+		
+		
+		// fail silently to prevent probing
+		return new ResponseEntity<Object>(entity, HttpStatus.OK);
```
The comment seems awkward. Since the code below happens on success and failures.

shinwachi commented:
Ok, changed code comments and format to address Ben's concerns.

shinwachi commented:
tnx
